File: elgamal.py
import random
from sympy import isprime

def generate_keys(bit_length):
    # Generate a prime number q of bit_length - 1
    prime_q = 1
    while not isprime(prime_q):
        prime_q = random.getrandbits(bit_length - 1)

    # Generate a safe prime p = 2q + 1
    safe_prime_p = 2 * prime_q + 1
    while not isprime(safe_prime_p):
        prime_q = random.getrandbits(bit_length - 1)
        safe_prime_p = 2 * prime_q + 1
    print(f"Found safe prime: {safe_prime_p}")

    # Generate a generator g of the group
    generator_g = 2
    while pow(generator_g, prime_q, safe_prime_p) == 1 or pow(generator_g, 2, safe_prime_p) == 1:
        generator_g = random.randint(2, safe_prime_p - 2)
    print(f"Found generator: {generator_g}")

    # Generate private and public keys
    private_key_a = random.randint(1, prime_q)  # private key
    public_key_h = pow(generator_g, private_key_a, safe_prime_p)  # public key

    return (safe_prime_p, generator_g, public_key_h), private_key_a

def encrypt(public_key, message):
    safe_prime_p, generator_g, public_key_h = public_key
    # Generate a random number r
    random_num_r = random.randint(0, safe_prime_p - 2)
    # Compute s = g^r mod p
    ciphertext_s = pow(generator_g, random_num_r, safe_prime_p)
    # Compute t = m*h^r mod p
    ciphertext_t = (message * pow(public_key_h, random_num_r, safe_prime_p)) % safe_prime_p
    # Return the ciphertext (s,t)
    return ciphertext_s, ciphertext_t

def decrypt(private_key, ciphertext, public_key):
    ciphertext_s, ciphertext_t = ciphertext
    # Compute sa = s^a mod p
    decrypted_sa = pow(ciphertext_s, private_key, public_key[0])
    # Compute m = t/sa mod p
    decrypted_message = (ciphertext_t * pow(decrypted_sa, -1, public_key[0])) % public_key[0]
    # Return the decrypted message
    return decrypted_message

# Example usage:
public_keys, private_keys = generate_keys(100)
print(f"Public keys: {public_keys}")
print(f"Private key: {private_keys}")

ciphertext = encrypt(public_keys, 10)
print(f"Ciphertext: {ciphertext}")

plaintext = decrypt(private_keys, ciphertext, public_keys)
print(f"Decrypted message: {plaintext}")

File: elgamal.c
#include <stdio.h>
#include <stdlib.h>
#include <gmp.h>
#include <time.h>

// Function to generate keys
void generate_keys(mpz_t prime, mpz_t generator, mpz_t private_key, mpz_t public_key, gmp_randstate_t state) {
    mpz_urandomb(prime, state, 100); // Generate a random 100-bit number
    mpz_nextprime(prime, prime); // Find the next prime after prime

    mpz_set_ui(generator, 2); // Set generator to 2

    mpz_urandomb(private_key, state, 100); // Generate a random 100-bit number
    mpz_mod(private_key, private_key, prime); // Reduce private_key modulo prime

    mpz_powm(public_key, generator, private_key, prime); // Compute public_key = generator^private_key mod prime
}

// Function to encrypt a message
void encrypt(mpz_t ciphertext1, mpz_t ciphertext2, mpz_t message, mpz_t prime, mpz_t generator, mpz_t public_key, gmp_randstate_t state) {
    mpz_t ephemeral_key;
    mpz_init(ephemeral_key);

    mpz_urandomb(ephemeral_key, state, 100); // Generate a random 100-bit number
    mpz_mod(ephemeral_key, ephemeral_key, prime); // Reduce ephemeral_key modulo prime

    mpz_powm(ciphertext1, generator, ephemeral_key, prime); // Compute ciphertext1 = generator^ephemeral_key mod prime

    mpz_t shared_secret;
    mpz_init(shared_secret);
    mpz_powm(shared_secret, public_key, ephemeral_key, prime); // Compute shared_secret = public_key^ephemeral_key mod prime

    mpz_mul(ciphertext2, message, shared_secret);
    mpz_mod(ciphertext2, ciphertext2, prime); // Compute ciphertext2 = message * shared_secret mod prime

    mpz_clear(ephemeral_key);
    mpz_clear(shared_secret);
}

// Function to decrypt a message
void decrypt(mpz_t decrypted_message, mpz_t ciphertext1, mpz_t ciphertext2, mpz_t prime, mpz_t private_key) {
    mpz_t shared_secret;
    mpz_init(shared_secret);
    mpz_powm(shared_secret, ciphertext1, private_key, prime); // Compute shared_secret = ciphertext1^private_key mod prime

    mpz_invert(shared_secret, shared_secret, prime); // Compute shared_secret = shared_secret^-1 mod prime

    mpz_mul(decrypted_message, ciphertext2, shared_secret);
    mpz_mod(decrypted_message, decrypted_message, prime); // Compute decrypted_message = ciphertext2 * shared_secret mod prime

    mpz_clear(shared_secret);
}

int main() {
    gmp_randstate_t state;
    gmp_randinit_mt(state);
    gmp_randseed_ui(state, time(NULL)); // Seed the random number generator with the current time

    mpz_t prime, generator, private_key, public_key;
    mpz_inits(prime, generator, private_key, public_key, NULL);

    generate_keys(prime, generator, private_key, public_key, state);

    gmp_printf("Public key: prime = %Zd, generator = %Zd, public_key = %Zd\n", prime, generator, public_key);
    gmp_printf("Private key: private_key = %Zd\n", private_key);

    mpz_t message1, message2, message3;
    mpz_init_set_ui(message1, 123); // Set message1 to 123
    mpz_init_set_ui(message2, 456); // Set message2 to 456
    mpz_init_set_ui(message3, 789); // Set message3 to 789

    gmp_printf("Original messages: message1 = %Zd, message2 = %Zd, message3 = %Zd\n", message1, message2, message3);

    mpz_t ciphertext1_1, ciphertext2_1, ciphertext1_2, ciphertext2_2, ciphertext1_3, ciphertext2_3;
    mpz_inits(ciphertext1_1, ciphertext2_1, ciphertext1_2, ciphertext2_2, ciphertext1_3, ciphertext2_3, NULL);

    encrypt(ciphertext1_1, ciphertext2_1, message1, prime, generator, public_key, state);
    encrypt(ciphertext1_2, ciphertext2_2, message2, prime, generator, public_key, state);
    encrypt(ciphertext1_3, ciphertext2_3, message3, prime, generator, public_key, state);

    gmp_printf("Encrypted messages: ciphertext1_1 = %Zd, ciphertext2_1 = %Zd, ciphertext1_2 = %Zd, ciphertext2_2 = %Zd, ciphertext1_3 = %Zd, ciphertext2_3 = %Zd\n", ciphertext1_1, ciphertext2_1, ciphertext1_2, ciphertext2_2, ciphertext1_3, ciphertext2_3);

    mpz_t decrypted_message1, decrypted_message2, decrypted_message3;
    mpz_inits(decrypted_message1, decrypted_message2, decrypted_message3, NULL);

    decrypt(decrypted_message1, ciphertext1_1, ciphertext2_1, prime, private_key);
    decrypt(decrypted_message2, ciphertext1_2, ciphertext2_2, prime, private_key);
    decrypt(decrypted_message3, ciphertext1_3, ciphertext2_3, prime, private_key);

    gmp_printf("Decrypted messages: message1 = %Zd, message2 = %Zd, message3 = %Zd\n", decrypted_message1, decrypted_message2, decrypted_message3);

    mpz_clears(prime, generator, private_key, public_key, message1, message2, message3, ciphertext1_1, ciphertext2_1, ciphertext1_2, ciphertext2_2, ciphertext1_3, ciphertext2_3, decrypted_message1, decrypted_message2, decrypted_message3, NULL);
    gmp_randclear(state);

    return 0;
}
